cli/src/main/java/org/aya/cli/library/incremental/InMemoryCompilerAdvisor.java

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.cli.library.incremental;
+
+import kala.collection.immutable.ImmutableSeq;
+import kala.collection.mutable.MutableMap;
+import org.aya.cli.library.source.LibraryOwner;
+import org.aya.cli.library.source.LibrarySource;
+import org.aya.core.def.GenericDef;
+import org.aya.core.serde.SerTerm;
+import org.aya.core.serde.Serializer;
+import org.aya.resolve.ResolveInfo;
+import org.aya.resolve.module.ModuleLoader;
+import org.aya.util.reporter.Reporter;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+
+public record InMemoryCompilerAdvisor(
+  @NotNull MutableMap<Path, FileTime> coreTimestamp,
+  @NotNull MutableMap<ImmutableSeq<String>, ResolveInfo> compiledCore
+) implements CompilerAdvisor {
+  public InMemoryCompilerAdvisor() {
+    this(MutableMap.create(), MutableMap.create());
+  }
+
+  @Override public boolean isSourceModified(@NotNull LibrarySource source) {
+    var coreLastModified = coreTimestamp.getOption(source.file());
+    try {
+      if (coreLastModified.isEmpty()) return true;
+      return Files.getLastModifiedTime(source.file())
+        .compareTo(coreLastModified.get()) > 0;
+    } catch (IOException ignore) {
+      return true;
+    }
+  }
+
+  @Override public void updateLastModified(@NotNull LibrarySource source) {
+    try {
+      coreTimestamp.put(source.file(), Files.getLastModifiedTime(source.file()));
+    } catch (IOException ignore) {
+    }
+  }
+
+  @Override public void prepareLibraryOutput(@NotNull LibraryOwner owner) {
+  }
+
+  @Override public void prepareModuleOutput(@NotNull LibrarySource source) {
+    // TODO: what if module name clashes?
```

Comment:
Return `bool`?

Reply:
Module name clashing should be addressed in shallow resolver. If addressed we don't need to change anything here. I guess it's ok to return `void`